db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/fk/Graph.java

```diff
@@ -0,0 +1,181 @@
+                EdgeNode.Type edgeType = getEdgeType(p.get(i-1), p.get(i));
+                if(edgeType != EdgeNode.Type.C) {
+                    addSurrogate(p.get(i-1), p.get(i), edgeType);
+                    // remove this edge
+                    removeEdge(p.get(i-1), p.get(i));
+                    // add the new edge again, but check again for cycles!
+                    addEdgeInternal(fromIdx, toIdx, type);
+                    return;
+                }
+            }
+        }
+        // cycle is unbreakable, bail out
+        StringBuilder sb = new StringBuilder();
+        sb.append("adding the constraint between ").append(
+                getName(p.get(0))).append(" and ").append(getName(p.get(p.size()-1))).append(" would cause the following illegal delete action cascade cycle");
+        for(int v : p) {
+            sb.append(" ").append(getName(v));
+        }
+        throw StandardException.newException(SQLState.LANG_DELETE_RULE_VIOLATION,
+                                             newConstraintName,
+                                             sb.toString());
+    }
+
+    public String getName(int index) {
+        for (Map.Entry<String, Integer> entry : vertexIndex.entrySet()) {
+            if (entry.getValue() == index) {
+                return entry.getKey();
+            }
+        }
+        throw new IllegalArgumentException("could not find name for item at the index: " + index);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("digraph {").append("\n");
+        int i = 0;
+        for (EdgeNode edge : edgeNodes) {
+            EdgeNode next = edge;
+            while (next != null) {
+                sb.append("\t").append(getName(i)).append(" -> ").append(getName(next.y)).append("[label=\"").append(next.type.toString()).append("\"];").append("\n");
+                next = next.next;
+            }
+            i++;
+        }
+        sb.append("}").append("\n");
+        return sb.toString();
+    }
+
+    public EdgeNode getEdge(int v) {
+        return edgeNodes.get(v);
+    }
+
+    public int getVertexCount() {
+        return vertexIndex.size();
+
+    }
+
+    public EdgeNode.Type getEdgeType(int from, int to) {
+        EdgeNode edgeNode = getEdge(from);
+        while(true) {
+            if(edgeNode == null) {
+                throw new IllegalArgumentException("no edge between " + getName(from) + " and " + getName(to));
+            }
+            if(edgeNode.y == to) {
+                return edgeNode.type;
+            }
+            edgeNode = edgeNode.next;
+        }
+    }
+
+    public void removeEdge(int from, int to) {
+        EdgeNode edgeNode = getEdge(from);
+        if(edgeNode == null) {
+            throw new IllegalArgumentException("no edge between " + getName(from) + " and " + getName(to));
+        }
+        if(edgeNode.y == to) {
+            edgeNodes.set(from, null);
```

Comment:
if it's the first edge, but there are other edges, this code would delete all edges, but i guess it should only delete the first one ->
`edgeNodes.set(from, edgeNodes.next);`

Reply:
Great point, thanks for spotting this, fixed it.